serverside_challenge_1/challenges/toru_shimizu/app/services/prices_service.rb

```diff
@@ -0,0 +1,83 @@
+# frozen_string_literal: true
+
+class PricesService
+  def create_response(ampere, volume, companies)
+    companies.map.with_index do |company, _index|
+      company['plans'].map do |plan|
+        basic_price = caluculate_basic(ampere, plan['basics'])
+        volume_price = caluculate_volume(volume, plan['volumes'])
+
+        # 契約可能なアンペア数の場合のみレスポンスに含める
+        if basic_price
+          { 'provider_name' => company['name'], 'plan_name' => company['plans'][0]['name'],
+            'price' => basic_price + volume_price }
+        end
+      end
+    end.flatten.compact
+  end
+
+  def caluculate_basic(ampere, basics)
+    plan = basics.find do |basic|
+      basic['ampere'].to_i == ampere
+    end
+    # 契約可能なアンペア数でない場合はnilを返す
+    plan ? plan['price'].to_i : nil
+  end
+
+  def caluculate_volume(value, volumes)
+    price = volumes.reduce(0) do |sum, volume|
+      max = volume['max'].to_i
+      min = volume['min'].to_i
+      price = volume['price'].to_f
+
+      if min <= value
+        # max以下の場合はvalue - minをpriceにかける
+        if max >= value
+          minus = value - min
+          sum += minus * price
+        end
+
+        # maxを超えている場合はmax - minをpriceにかける
+        if max < value
+          minus = max - min
+          sum += minus * price
+        end
+
+        # 上限値が設定されていない場合。一律料金の場合もここに到達する
+        if max.zero?
+          vol = value - max
+          sum += vol * price
```

Comment:
maxが0なら下記で良さそうですね。

```suggestion
          sum += volume * price
```

またmaxはyamlではnullで登録していて、to_iメソッドで0になっていると思うのですが、yaml内に-infで登録し、to_iとしているところto_fやbigdecimal型を使うように変更すると、35行目のifがtrueになるので、47 ~ 50行目の処理がまるっと不要になるように見えます。